### ISSUE TITLE
How to specify positions manually?

### ISSUE THREAD

**Issue by marton-avrios** [google-research/text-to-text-transfer-transformer#325]
I want to specify positions for tokens instead of using the order of tokens. I tried to add "inputs_position" but it seems to be ignored.

**Comment by craffel** [comment]
This is not currently supported; the positions are added manually by the packing and padding code https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/dataset.py#L22
PRs welcome.

**Comment by marton-avrios** [comment]
In theory is it possible to specify the same position for 2 different tokens? Or will it break relative attention?

**Comment by craffel** [comment]
Relative attention should just compute a (learned) logit bias according to the "distance" between the two positions, so there should be no issue with duplicate positions (off the top of my head, anyways).

**Comment by marton-avrios** [comment]
and this also means that positions can even be factions? like 12.5, 34.7, etc.? even greater than maximum sequence length or even smaller than 0?

**Comment by craffel** [comment]
Sure, why not! Here is the code that computes it, for reference:
https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/transformer_layers.py#L279

**Comment by marton-avrios** [comment]
I replaced `tf.range(new_seq_len, dtype=tf.int32)` with `tf.range(new_seq_len, dtype=tf.int32) * 3 + 7` in the pack or pad code as default input positions so they should be like 7, 10, 13, etc. Then I `mtf.Print`-ed `relative_attention` in `compute_bias` but it is still `[[0 -1 -2 -3 -4 -5 -6 -7 -8 -9][1 0 -1 -2 -3 -4 -5 -6 -7 -8][2...]...]` (my maximum input and target length is 10 for this experiment). Should not they be [[0, -3, -6, -9...], [3, 0, -3, -6],...]?

**Comment by craffel** [comment]
Do you mean `relative_position`? https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/transformer_layers.py#L294 Are `memory_position` and `context.position` not what you intended?

**Comment by marton-avrios** [comment]
Sorry, yes, I mean `relative_position`. Just after the line you referenced I put this: `relative_position = mtf.Print(relative_position, [relative_position])` and the output is like I mentioned. Also in this line https://github.com/tensorflow/mesh/blob/efcd89f66d53ffddc5d13d37041d774d1bf7d210/mesh_tensorflow/transformer/dataset.py#L645 I replaced `tf.range(new_seq_len, dtype=tf.int32)` with `tf.range(new_seq_len, dtype=tf.int32) * 3 + 7`.

**Comment by craffel** [comment]
Do you have `pack_dataset.use_custom_ops` set to `True`, either via a caller or via gin? https://github.com/tensorflow/mesh/blob/efcd89f66d53ffddc5d13d37041d774d1bf7d210/mesh_tensorflow/transformer/dataset.py#L474 If so, you won't end up using the code snippet you modified, you'll use the custom TF op instead for packing. The code snipped you modified also won't be triggered unless you are packing (and, for example, during eval we pad and don't pack).

**Comment by marton-avrios** [comment]
Nope, `pack_dataset.use_custom:ops` is `False`. However I found this: https://github.com/tensorflow/mesh/blob/d353aa9ff6644048bfaec737b6fb5fada03085e2/mesh_tensorflow/transformer/transformer.py#L953 This is responsible for disregarding any custom position information I think.

**Comment by marton-avrios** [comment]
...yes, it was. I commented out that part and now I get the expected `relative_position` printed so now I can input custom positions.
